Feature: Server info

wutschel opened the pull request:
## Description
<!--- Detailed info for reviewers and developers -->
Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/389.

This PR implements a detailed server info, comparable to Kodi's System Info. The current state is shown in the screenshot below. More details can be shown, if this should be required. I am avoiding to add eye-candy like bars for memory consumption or the CPU load as I would need to parse Kodi's responses, and I am already seeing different string formats between my main test systems 17.6 and 19.3.

Screenshots (Kodi 19.3 on a VM):
<a href="https://abload.de/image.php?img=bildschirmfoto2021-121tk7s.png"><img src="https://abload.de/img/bildschirmfoto2021-121tk7s.png" /></a>

## Summary for release notes
<!--- This will be shown in Testflight to end users / testers -->
<!--- Please fill in, as usually PR title isn't clear to ordinary users -->
<!--- If your changes don't modify app files, please add prefix [not app] -->
Feature: Server info

wutschel commented:
@kambala-decapitator, you feel this is a useful feature?

kambala-decapitator commented:
looks good, wouldn’t hurt :)

Do you plan to add auto refresh?

wutschel commented:
> Do you plan to add auto refresh?

I didn't look into this, but I guess if this feature is added, I should  do so.

wutschel commented:
Added a timer to update the server info view. But I guess I overdid the invalidation of the timer. Does it make sense to  invalidate in both `viewWillDisappear` and `viewDidDisappear`?

kambala-decapitator commented:
> Added a timer to update the server info view. But I guess I overdid the invalidation of the timer. Does it make sense to  invalidate in both `viewWillDisappear` and `viewDidDisappear`?

It's sufficient to do only in one of them, usually "did" is fine (but might depend on the use case).

But invalidating it in dealloc makes no sense because it will never happen: timer holds a strong reference to its target essentially causing a retain cycle in your case.

wutschel commented:
> It's sufficient to do only in one of them, usually "did" is fine (but might depend on the use case).
> 
> But invalidating it in dealloc makes no sense because it will never happen: timer holds a strong reference to its target essentially causing a retain cycle in your case.

Perfect, applied the changes in a fixup commit.

kambala-decapitator commented on XBMC Remote/HostManagementViewController.m:
```diff
@@ -342,6 +350,110 @@ - (void)infoView {
+    if (!serverInfoView.hidden) {
+        [self updateServerInfo];
+        // Start timer to update the server info view
+        [serverInfoTimer invalidate];
```
Seems you can invalidate it unconditionally?

wutschel replied:
True, will change this.

kambala-decapitator commented on XBMC Remote/HostManagementViewController.m:
```diff
@@ -342,6 +350,110 @@ - (void)infoView {
+    [string1 addAttribute:NSForegroundColorAttributeName value:UIColor.lightGrayColor range:NSMakeRange(0, string1.length)];
+    NSDictionary *normalFontAttrib = @{NSFontAttributeName: [UIFont systemFontOfSize:fontSize]};
+    NSMutableAttributedString *string2 = [[NSMutableAttributedString alloc] initWithString:text attributes:normalFontAttrib];
+    [string2 addAttribute:NSForegroundColorAttributeName value:UIColor.whiteColor range:NSMakeRange(0, string2.length)];
```
Why not providing it in the attribute dictionary directly?

wutschel replied:
Like this?

```
NSDictionary *boldFontAttrib = @{
    NSFontAttributeName: [UIFont boldSystemFontOfSize:fontSize],
    NSForegroundColorAttributeName: UIColor.lightGrayColor
};
NSMutableAttributedString *string1 = [[NSMutableAttributedString alloc] initWithString:name attributes:boldFontAttrib];
```

kambala-decapitator replied:
Yes

wutschel replied:
Ok, done.

kambala-decapitator commented on XBMC Remote/HostManagementViewController.m:
```diff
@@ -433,6 +566,14 @@ - (void)viewDidLoad {
     [addHostButton setTitleColor:UIColor.whiteColor forState:UIControlStateSelected];
     addHostButton.titleLabel.shadowOffset = CGSizeZero;
     
+    [serverInfoButton setBackgroundImage:[UIImage new] forState:UIControlStateNormal];
+    [serverInfoButton setBackgroundImage:[UIImage new] forState:UIControlStateHighlighted];
+    [serverInfoButton setBackgroundImage:[UIImage new] forState:UIControlStateSelected];
```
Not clear why this is needed

wutschel replied:
Oh, this came from copy-pasting the other button declarations in this controller. Obviously the others also do not need this clean-up as there is never a background set. Will add a fixup commit for the newly added button and another commit to remove the unneeded calls of setBackgroundImage.

kambala-decapitator reviewed (changes requested):
Please use UITextView in readonly mode instead to make things easier ;)

kambala-decapitator commented:
new commits look fine.

> Please use UITextView in readonly mode instead to make things easier ;)

@wutschel will you do this?

wutschel commented:
> @wutschel will you do this?

Can you give me a hint what you are looking for?

kambala-decapitator commented:
> > @wutschel will you do this?
> 
> Can you give me a hint what you are looking for?

You're adding tons of labels onto scroll view. `UITextView` is exactly a "scrollable label".

wutschel commented:
Understood. I now tried to directly use `UITextView` and set the attributed string with all content. But I cannot get the string displayed, the view itself with the background color shows up. Any idea where to look at?

Btw, what works is to keep using `UIScrollView` and only add a single `UILabel` with all content to it.

kambala-decapitator commented:
> Understood. I now tried to directly use `UITextView` and set the attributed string with all content. But I cannot get the string displayed, the view itself with the background color shows up. Any idea where to look at?

sounds strange. Do you experience the same in a clean test project?

wutschel commented:
Hah, found it. I had added a loop which removes all subviews before adding the new ones. This also removed the text :) Works now. Will tweak a little and then provide a reworked version.

wutschel commented:
Here we go.

kambala-decapitator commented on XBMC Remote/HostManagementViewController.m:
```diff
@@ -422,37 +423,25 @@ - (void)showServerInfoView {
 }
 
-- (CGFloat)addLabel:(NSString*)name text:(NSString*)text offset:(CGFloat)offset {
+- (NSMutableAttributedString*)formatInfo:(NSString*)name text:(NSString*)text {
```
don't see a reason to return mutable object

wutschel replied:
Right.

kambala-decapitator commented on XBMC Remote/HostManagementViewController.m:
```diff
@@ -379,34 +379,35 @@ - (void)updateServerInfo {
-            offset = [self addLabel:@"GPU Temp" text:methodResult[@"System.GPUTemperature"] offset:offset];
-            offset = [self addLabel:@"HDD Temp" text:methodResult[@"System.HddTemperature"] offset:offset] + BLOCK_MARGIN;
+            NSMutableAttributedString *infoString = [[NSMutableAttributedString alloc] initWithString:@""];
+            NSMutableAttributedString *newLine = [[NSMutableAttributedString alloc] initWithString:@"\n"];
```
why is it mutable?

wutschel replied:
Copy pasted, removed this.

kambala-decapitator commented on XBMC Remote/HostManagementViewController.m:
```diff
@@ -379,34 +379,35 @@ - (void)updateServerInfo {
-            offset = [self addLabel:@"CPU Temp" text:methodResult[@"System.CPUTemperature"] offset:offset];
-            offset = [self addLabel:@"GPU Temp" text:methodResult[@"System.GPUTemperature"] offset:offset];
-            offset = [self addLabel:@"HDD Temp" text:methodResult[@"System.HddTemperature"] offset:offset] + BLOCK_MARGIN;
+            NSMutableAttributedString *infoString = [[NSMutableAttributedString alloc] initWithString:@""];
```
can be simplified to `[NSMutableAttributedString new]`

wutschel replied:
Right.

kambala-decapitator reviewed (changes requested):
much better now!

wutschel commented:
> much better now!

True, looks much cleaner now. Thanks for nudging ;)

kambala-decapitator commented on XBMC Remote/HostManagementViewController.m:
```diff
@@ -380,7 +380,7 @@ - (void)updateServerInfo {
         if (error == nil && methodError == nil) {
             NSMutableAttributedString *infoString = [NSMutableAttributedString new];
-            NSMutableAttributedString *newLine = [[NSMutableAttributedString alloc] initWithString:@"\n"];
+            NSAttributedString *newLine = [[NSMutableAttributedString alloc] initWithString:@"\n"];
```
```suggestion
            NSAttributedString *newLine = [[NSAttributedString alloc] initWithString:@"\n"];
```

wutschel replied:
Thanks, will apply this in the update.

wutschel commented:
Squashed the fixup commits into a single commit.
